### PR TITLE
Compliance Controller Improvements

### PR DESCRIPTION
**Pull request by AdamPettinger**
This is an update to this pull request to kinetic-devel: [(WIP) New type of compliance controller #393](https://github.com/ros-controls/ros_controllers/pull/393).

Updated to Melodic, added singularity checking, and improved directionally constrained compliance.

Tested on UR5 with ATI Gamma Force/Torque sensor. Videos showing improved directional control:

1. [Compliant only in EEF Forward and EEF roll directions](https://drive.google.com/file/d/1i77SVG8YFoIGYAzKRylypvYj8J8b838f/view?usp=sharing)
2. [Grasping a valve with compliance](https://drive.google.com/file/d/1FAsqDhwRbRmkkmAsjUnJbrf_N2DZzJQ1/view?usp=sharing) and no other user input

### CODE REVIEW

**Comment by bmagyar** [comment]
Thanks for this guys. Would it be possible to squash/rebase this PR to reduce the number of commits (and potential conflicts)?

File: compliance_controller/README.md [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+### Building
+
+* sudo apt install ros-kinetic-ros-contro*
+
+* Clone universal_robot (kinetic-devel branch), https://github.com/ros-industrial/universal_robot
+
+* Clone ur_modern_driver (kinetic-devel branch), https://github.com/ros-industrial/ur_modern_driver
+
+* Clone NRG's ros_controllers repo (kinetic-compliance branch), https://github.com/UTNuclearRobotics/ros_controllers
```
Comment (bmagyar) [inline compliance_controller/README.md:13]: Needs an update?

File: compliance_controller/compliance_control_msgs/package.xml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0"?>
+<package format="2">
+  <name>compliance_control_msgs</name>
+  <version>0.0.0</version>
+  <description>The compliance_control_msgs package</description>
+
+  <maintainer email="andyz@utexas.edu">AndyZe</maintainer>
```
Comment (bmagyar) [inline compliance_controller/compliance_control_msgs/package.xml:7]: Missing author

File: compliance_controller/wrench_to_joint_vel_pub/include/wrench_to_joint_vel_pub/compliant_control.h [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+namespace wrench_to_joint_vel_pub
+{
+static const std::string LOGNAME = "compliant_control";
+/**
+ * Dimension enum.
+ */
+enum Dimension
+{
+  NUM_DIMS = 6  // 3 translational, 3 rotational dimensions
```
Comment (bmagyar) [inline compliance_controller/wrench_to_joint_vel_pub/include/wrench_to_joint_vel_pub/compliant_control.h:67]: How about a single cost instead?

File: compliance_controller/wrench_to_joint_vel_pub/src/wrench_to_joint_vel_pub/wrench_to_joint_vel_pub.cpp [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+    res.success = false;
+    return false;
+  }
+}
+
+bool PublishCompliantJointVelocities::adjustDamping(
+    compliance_control_msgs::AdjustDamping::Request& req,
+    compliance_control_msgs::AdjustDamping::Response& res)
+{
+  std::vector<double> damping = req.damping.data;
```
Comment (bmagyar) [inline compliance_controller/wrench_to_joint_vel_pub/src/wrench_to_joint_vel_pub/wrench_to_joint_vel_pub.cpp:168]: Const

File: velocity_controllers/CHANGELOG.rst [hunk 5]
```diff
@@ -21,6 +21,10 @@ Changelog for package velocity_controllers
 -------------------
 
 0.14.0 (2018-04-27)
+0.13.4 (2018-06-26)
```
Comment (bmagyar) [inline velocity_controllers/CHANGELOG.rst:24]: Don't make changes to this file please

File: traj_or_jog_controller/include/traj_or_jog_controller/traj_or_jog_controller.h [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+  }
+
+  /**
+   * \brief Override the callback for the JointTrajectoryController action server.
+   */
+  void goalCB(GoalHandle gh);
+};
+}
+
+#endif  // header guard
```
Comment (bmagyar) [inline traj_or_jog_controller/include/traj_or_jog_controller/traj_or_jog_controller.h:90]: On melodic you have C++14 by default, this archaic tech can now be replaced by `#pragma once`

File: ros_controllers/CHANGELOG.rst [hunk 7]
```diff
@@ -19,6 +19,10 @@ Changelog for package ros_controllers
 -------------------
 
 0.14.0 (2018-04-27)
+0.13.4 (2018-06-26)
+-------------------
+
```
Comment (bmagyar) [inline ros_controllers/CHANGELOG.rst:24]: No meddling with changelogs please

**Review by bmagyar — CHANGES_REQUESTED** [review]
First of all, thanks for contributing!! 

Please ensure that you only change code in the files you are adding. Specifically changelogs shouldn't be touched. There's a new Git modules file added, please get rid of it. Make sure you are listed as author and maintainer in each package and please add me as maintainer as well.

I haven't finished with the review but wanted to post these comments so you have some feedback.

File: compliance_controller/compliance_controller/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+)
+
+include_directories(include ${catkin_INCLUDE_DIRS} ${traj_or_jog_controller_INCLUDE_DIRS} ${Boost_INCLUDE_DIR})
+
+add_library(compliance_controller src/compliance_controller.cpp)
+target_link_libraries(compliance_controller ${catkin_LIBRARIES})
+
+install(DIRECTORY include/${PROJECT_NAME}/
+  DESTINATION ${CATKIN_PACKAGE_INCLUDE_DESTINATION})
+
```
Comment (bmagyar) [inline compliance_controller/compliance_controller/CMakeLists.txt:35]: We proud ourselves here for our tests. They ensure that continued improvements to the controllers and the ros_control framework keep everything working. Could you please add some rostests that ensure some basic functionality? Look at joint traj controller or diff drive controller for good examples.

**Review by bmagyar — CHANGES_REQUESTED** [review]

**Comment by AdamPettinger** [comment]
Thanks for the feedback, I am making the requested changes. In terms of squashing it down, I am struggling to do so. This is my first big PR and unfortunately is a messy one :(. A lot of the early commits are not related to the compliance controller and are causing me problems.

Locally I was able to start with a fresh `ros-controls:melodic-devel` branch, and cherry pick the commits for the compliance controller. Then I rebased to squash it to 1-2 commits and everything looks good (only files changed are the ones we added, builds and runs as expected). At this point I am not sure if it is better to merge this branch with our `UTNuclearRobotics:melodic-devel` branch and force push it to update this PR, or to close the PR and open a new one, or to do something else entirely. Do you have insights here?

As for actual code changes, I will work to incorporate the changes requested here, as well as in the original PR for the kinetic branch

**Comment by bmagyar** [comment]
First of all I suggest backing it all up. Avoid force pushing to your main branch, create a separate one for this MR and if you want to merge the end result back once its emerged here, so be it. 
Github allows for retargeting PRs including the source if I remember well... 

You should be able to do an interactive rebase. I suggest to do it in a couple of steps. First, review the commits that show up on this PR and drop all that don't belong to your team. This will eliminate the changelog changes and other artifacts of releasing. Then as a second step see if there are hacky commits (the message sometimes tells you) and drop those too. Now, try squashing the first few commits and move on from here. If you have massive conflicts, you may want to rethink your strategy and see if the commit causing it can be eliminated first.

**Comment by AdamPettinger** [comment]
Closed this PR to open a new, cleaner one with proper rebasing, some additional features, and some of the fixes requested here

**Comment by bmagyar** [comment]
Thank you!

**Pull request closed without merging**